File: test/benchmark/beng/js/fannkuch.js
// BENG Benchmark: fannkuch-redux (Node.js reference)

const N = parseInt(process.argv[2] || "7");

function fannkuch(n) {
    const perm = new Int32Array(n);
    const perm1 = new Int32Array(n);
    const count = new Int32Array(n);
    let maxFlips = 0, checksum = 0, permCount = 0;

    for (let i = 0; i < n; i++) perm1[i] = i;

    let r = n;
    while (true) {
        while (r !== 1) { count[r - 1] = r; r--; }

        for (let i = 0; i < n; i++) perm[i] = perm1[i];

        let flips = 0;
        let k = perm[0];
        while (k !== 0) {
            let lo = 0, hi = k;
            while (lo < hi) {
                const tmp = perm[lo]; perm[lo] = perm[hi]; perm[hi] = tmp;
                lo++; hi--;
            }
            flips++;
            k = perm[0];
        }

        if (flips > maxFlips) maxFlips = flips;
        checksum += (permCount % 2 === 0) ? flips : -flips;
        permCount++;

        // next permutation
        let found = false;
        r = 1;
        while (r < n) {
            const perm0 = perm1[0];
            for (let i = 0; i < r; i++) perm1[i] = perm1[i + 1];
            perm1[r] = perm0;
            count[r]--;
            if (count[r] > 0) { found = true; break; }
            r++;
        }
        if (!found) break;
    }

    console.log(checksum);
    console.log(`Pfannkuchen(${n}) = ${maxFlips}`);
}

fannkuch(N);

File: test/benchmark/beng/python/fannkuch.py
#!/usr/bin/env python3
# BENG Benchmark: fannkuch-redux (Python)
import sys
import time

N = int(sys.argv[1]) if len(sys.argv) > 1 else 7


def fannkuch(n):
    perm = list(range(n))
    perm1 = list(range(n))
    count = [0] * n
    max_flips = 0
    checksum = 0
    perm_count = 0

    r = n
    while True:
        while r != 1:
            count[r - 1] = r
            r -= 1

        for i in range(n):
            perm[i] = perm1[i]

        flips = 0
        k = perm[0]
        while k != 0:
            lo, hi = 0, k
            while lo < hi:
                perm[lo], perm[hi] = perm[hi], perm[lo]
                lo += 1
                hi -= 1
            flips += 1
            k = perm[0]

        if flips > max_flips:
            max_flips = flips
        checksum += flips if perm_count % 2 == 0 else -flips
        perm_count += 1

        # next permutation
        found = False
        r = 1
        while r < n:
            perm0 = perm1[0]
            for i in range(r):
                perm1[i] = perm1[i + 1]
            perm1[r] = perm0
            count[r] -= 1
            if count[r] > 0:
                found = True
                break
            r += 1
        if not found:
            break

    return checksum, max_flips


t0 = time.perf_counter_ns()
checksum, max_flips = fannkuch(N)
t1 = time.perf_counter_ns()
print(checksum)
print(f"Pfannkuchen({N}) = {max_flips}")
print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")
